3.Linear_Modeling/nbconverted/fit_linear_model.py

```diff
@@ -0,0 +1,178 @@
+cell_count_df = (
+    df.groupby("Metadata_Well")["Metadata_Plate"]
+    .count()
+    .reset_index()
+    .rename(columns={"Metadata_Plate": "Metadata_number_of_singlecells"})
+)
+
+df = df.merge(cell_count_df, on="Metadata_Well")
+
+
+# Define CellProfiler features
+cp_features = infer_cp_features(df)
+
+print(f"We are testing {len(cp_features)} CellProfiler features")
+# Drop na and reindex accordingly
+df = df.dropna()
+df = df.reindex()
+df = df.assign(
+    Metadata_Treatment_and_Dose=lambda x: df["Metadata_treatment"]
+    + "_"
+    + df["Metadata_dose"]
+)
+df["Metadata_Treatment_and_Dose"].unique()
+
+
+# In[5]:
+
+
+treatment = [
+    "LPS_10µg/ml",
+    "Disulfiram_2.5µM",
+    "LPS_1µg/ml",
+    "Disulfiram_0.1µM",
+    "H2O2_500µM",
+    "Thapsi_10µM",
+    "H2O2_50µM",
+    "Thapsi_1µM",
+    "ATP_1mM",
+    "LPS + Nigericin_1µg/ml + 10µM",
+    "ATP_0.1mM",
+    "LPS + Nigericin_1µg/ml + 1µM",
+    "Flagellin_1µg/ml",
+    "Flagellin_0.1µg/ml",
+]
+control = ["DMSO 0.1%_0", "Media only_0"]
+
+
+def feature_importance_linear_model(
+    df: pd.DataFrame, treatment_list: list, control_list: list
+):
+    dosage_treatments_list = []
+    for i in treatment_list:
+        dosage_treatments_list.append(i)
+    for i in control_list:
+        dosage_treatments_list.append(i)
+    df = df.query("Metadata_Treatment_and_Dose == @dosage_treatments_list")
+    # Add dummy matrix of categorical genotypes
+    treatment = pd.get_dummies(data=df.Metadata_Treatment_and_Dose)
+    # Setup linear modeling framework
+    variables = ["Metadata_number_of_singlecells"]
+    X = df.loc[:, variables]
+
+    X = pd.concat([X, treatment], axis=1)
+
+    columns_list = []
+    columns_list.append("feature")
+    columns_list.append("r2score")
+
+    for i in X:
+        columns_list.append(i)
+
+    # Fit linear model for each feature
+    lm_results = []
+    for cp_feature in cp_features:
+        # Subset CP data to each individual feature (univariate test)
+        cp_subset_df = df.loc[:, cp_feature]
+
+        # Fit linear model
+        lm = LinearRegression(fit_intercept=True)
+        lm_result = lm.fit(X=X, y=cp_subset_df)
+
+        # Extract Beta coefficients
+        # (contribution of feature to X covariates)
+        coef = lm_result.coef_
+
+        # Estimate fit (R^2)
+        r2_score = lm.score(X=X, y=cp_subset_df)
+
+        # Add results to a growing list
+        lm_results.append([cp_feature, r2_score] + list(coef))
+
+        # Convert results to a pandas DataFrame
+    lm_results = pd.DataFrame(lm_results, columns=columns_list)
+
+    # Output file
+    # lm_results.to_csv(output_cp_file, sep="\t", index=False)
+
+    print(lm_results.shape)
+    lm_results.head()
+    return lm_results
+
+
+lm_results = feature_importance_linear_model(df, treatment, control)
+
+
+# In[6]:
+
+
+def plot_lm(lm_df: pd.DataFrame, x: str, y: str, fill: str):
```

Comment:
this probably also doesn't need to be a function - but maybe! (if so, please define it immediately after the import statements, or, in a separate `.py` file that you import)